src/main/java/org/apache/sling/feature/cpconverter/handlers/slinginitialcontent/VaultContentXMLContentCreator.java

```diff
@@ -0,0 +1,228 @@
+        this.repositoryPath = repositoryPath;
+        this.targetOutputStream = targetOutputStream;
+        this.packageAssembler = packageAssembler;
+        this.namespaceRegistry = namespaceRegistry;
+        this.isFileDescriptorEntry = isFileDescriptorEntry;
+    }
+
+    public void setIsXmlProcessed(){
+        this.xmlProcessed = true;
+    }
+    
+    @Override
+    public void createNode(String name, String primaryNodeType, String[] mixinNodeTypes) throws RepositoryException {
+        
+        final String elementName;
+        final String jcrNodeName; 
+        if(xmlProcessed && isFirstElement){
+            elementName = "jcr:root";
+            primaryNodeName = name;
+            jcrNodeName = name;
+            isFirstElement = false;
+        }else if(StringUtils.isNotBlank(name)){
+            elementName = getValidElementName(name);
+            jcrNodeName = name;
+        }else{
+            elementName = "jcr:root";
+            jcrNodeName = null;
+        }
+
+        
+        final String basePath;
+        if(parentNodePathStack.isEmpty()){
+            basePath = repositoryPath;
+        }else{
+            StringBuilder basePathBuilder = new StringBuilder(repositoryPath);
+            for(Iterator<XMLNode> xmlNodeIterator = parentNodePathStack.descendingIterator(); xmlNodeIterator.hasNext();){
+                XMLNode parent = xmlNodeIterator.next();
+                String parentJcrNodeName = parent.getJcrNodeName();
+                if(StringUtils.isNotBlank(parentJcrNodeName)){
+                    basePathBuilder.append("/");
+                    basePathBuilder.append(parentJcrNodeName);
+                }
+            }
+            basePath = basePathBuilder.toString();
+        }
+
+        //if we are dealing with a descriptor file, we should use nt:file as default primaryType. 
+
+        String defaultNtType = isFileDescriptorEntry ? JcrConstants.NT_FILE : JcrConstants.NT_UNSTRUCTURED;
+        String toUsePrimaryNodeType = StringUtils.isNotBlank(primaryNodeType) ? primaryNodeType : defaultNtType;
+        XMLNode intermediateNode = new XMLNode(packageAssembler,basePath,elementName,jcrNodeName, toUsePrimaryNodeType, mixinNodeTypes);
+        //add the created node to the correct parent if present
+        if(currentNode != null){
+            currentNode.addChildNode(elementName, intermediateNode);
+        }
+        //switch the current node 
+        currentNode = intermediateNode;
+
+        if(ArrayUtils.isNotEmpty(mixinNodeTypes)){
+            currentNode.addProperty(JcrConstants.JCR_MIXINTYPES, "[" + String.join(",", mixinNodeTypes) + "]");
+        }
+
+        parentNodePathStack.push(currentNode);
+    }
+
+    private String getValidElementName(String name) {
+        if(StringUtils.isNumeric(name.substring(0,1))){
+            return "_" + name; 
+        }
+        return name;
+    }
+
+    public String getPrimaryNodeName() {
```

Comment:
same here.
in addition i would suggest to move this method together with setIsXmlProcessed and not have it in between methods that are defined by the interface.